streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -275,31 +259,70 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
                               final RecordConverter recordConverter) {
         for (final TopicPartition topicPartition : topicPartitions) {
             globalConsumer.assign(Collections.singletonList(topicPartition));
+            long offset;
             final Long checkpoint = checkpointFileCache.get(topicPartition);
             if (checkpoint != null) {
                 globalConsumer.seek(topicPartition, checkpoint);
+                offset = checkpoint;
             } else {
                 globalConsumer.seekToBeginning(Collections.singletonList(topicPartition));
+                final AtomicLong position = new AtomicLong();
+                retryUntilSuccessOrThrowOnTaskTimeout(
+                    () -> position.set(globalConsumer.position(topicPartition)),
+                    String.format(
+                        "Failed to get position for partition %s. The broker may be transiently unavailable at the moment.",
+                        topicPartition
+                    )
+                );
+                offset = position.get();
             }
 
-            long offset = globalConsumer.position(topicPartition);
             final Long highWatermark = highWatermarks.get(topicPartition);
             final RecordBatchingStateRestoreCallback stateRestoreAdapter =
                 StateRestoreCallbackAdapter.adapt(stateRestoreCallback);
 
             stateRestoreListener.onRestoreStart(topicPartition, storeName, offset, highWatermark);
             long restoreCount = 0L;
 
+            long deadlineMs = NO_DEADLINE;
             while (offset < highWatermark) {
                 try {
                     final ConsumerRecords<byte[], byte[]> records = globalConsumer.poll(pollTime);
+                    if (records.isEmpty()) {
+                        if (taskTimeoutMs == 0L) {
+                            deadlineMs = maybeUpdateDeadlineOrThrow(
```

Comment:
For `poll()`, even if retrying is disabled, we need to retry to give a fetch request (with default request timeout of 30 seconds but only a default `pollTime` of 100ms) a fair change to actually return.

At least I believe this makes sense. Also \cc @guozhangwang

Reply:
Man, this is confusing. I think I see what you're getting at, but it seems pretty strange to have to go to all that trouble to extract the consumer config so that we can apply a shorter timeout on each poll, but then loop around until the originally configured client timeout passes.

Can you explain how the outcome is different than just calling `globalConsumer.poll(requestTimeoutMs)`?

It also seems strange to extract a consumer timeout configuration that specifically does not apply to `poll` and apply it to `poll`. This seems like it would violate users' expectations when they set that configuration value.

Why wouldn't we instead apply the non-progress timeout (`taskTimeoutMs`), since it seems like that's exactly what it's for? I.e., it seems like `globalConsumer.poll(pollTime + taskTimeoutMs)` might be the most appropriate choice here?

Reply:
> Can you explain how the outcome is different than just calling globalConsumer.poll(requestTimeoutMs)?

I did consider it, but was not 100% sure if this might be better. Was also concerned about "miss using" request timeout (as you already mentioned) as we have `StreamsConfig.POLL_TIMEOUT_MS`.

I like the proposal to use `pollTime + taskTimeoutMs` though!

Reply:
Actually, we need to use all three: `pollTime + requestTimeout + taskTimeoutMs`

If we don't include `requestTimeout` and `taskTimeoutMs` is set to `0` (or a smaller value than `requestTimeout` in general), the first `poll()` would return no data with very high probability and we would fail what seems not desirable.

Reply:
Btw: just for completeness: we will still use `pollTime` unmodified in `globalConsumer.poll(...)` during regular processing, so the config has still value. Only in the startup bootstrapping phase, we would use the sum of poll/request/task-timeout to not fail unnecessarily.

Also note that during normal processing, we don't apply `taskTimeoutMs` atm. Not sure if we should, but I would not want to include it in this PR. -- It won't block the "main" processing if the global `poll()` loop would not make progress because the `StreamThreads` are already running, and thus it's impact is different -- also, during normal processing the global `highWatermark` may move and thus we need a different strategy anyway compared to the bootstrap phase.

Reply:
Ah, that last point sounds good.

I'm not opposed, but it still seems strange to grab a random timeout from the client config and apply it to the `poll` method. It kind of makes me wonder what we're really trying to do here. If we want to give poll at least 30 seconds to return data, then we can just give it at least 30 seconds, right? No reason to abuse the client configuration.

On the other hand, the pollTime may not even appropriate for the global thread, right? It seems more like it was meant for StreamThread to make sure we don't block too long and violate the max poll interval. But since the global thread is only assigned and not subscribed, it has no constraint on how long it should block, except for the taskTimeoutMs config, right?

Clearly, we need to set some kind of lower bound on it, though, but it's not clear that it needs to be configurable. Anyway, just food for thought. Feel free to keep the requestTimeout if you really think it's appropriate.

Reply:
Well, the requestTimeout is no really "random" -- if we send a fetch request, it's the maximum time until the producer would abort the request and retry. -- _If_ poll() would throw a TimeoutException, this would be the point when we get a timeout if there are no retries. -- To be fair, maybe using the consumer's _default_api_timeout_ms_ config might be better than using requestTimeout though? -- The point being is, that picking a consumer config seems to be the best thing we can do for this case instead of a hard-coded number (or introducing a new config for this specific corner case).

> pollTime may not even appropriate for the global thread

I think it does (even if differently). At least for regular processing: (1) we interleave polling and flushing, and also when we stop the client, we want it to responsive and not being blocked in poll() for 5 minutes (taskTimeoutMs default) for this case. (What of course implies that one cannot really "intercept" the startup phase atm -- something we might want to address, but maybe not in this PR.)

> Clearly, we need to set some kind of lower bound on it, though, but it's not clear that it needs to be configurable.

I think the worst we can do is to make it not configurable. :)

Reply:
Thanks, @mjsax , that's fair.

I promise I'm no equivocating here; I'm just trying to understand why my intuition is that this is wrong.

It seems like maybe the fundamental problem here is that we can't distinguish among a successful poll that returns no data, a failure to poll, and a pending async fetch request. The one thing we know is that the end offset is beyond our current position, so there _should_ be data to poll, so we can assume that an empty return means either that the fetch failed internally or it hasn't completed yet.

Stepping back, this seems to be related to the problem of task idling, in which it's pointless to "idle" for a time so short that we have no chance to actually get a response back from the broker.

I feel like this is substantially my fault from #4855 / KIP-266. The purpose of making this API completely async was to avoid harming liveness in situations where we might have a relatively strict deadline. But that's not the case here.

I guess the "poor man's" solution we're going for here is to block poll for at least long enough to allow for a complete fetch round-trip from the broker. If we know that there was a round-trip, and we didn't get any data, then we can conclude that there was an error (since we know there is data to get). Since we can't know that there was a round trip, we weaken the condition to: if we know it's been long enough that there should have been a round-trip and we don't get data, we conclude there was probably an error.

In your KIP, we specified we would start the task timer _after_ the first error, so it seems like we really want to just block the poll for the round-trip time, and then apply your "update deadline, etc." function. I'm with you now that to get the round-trip time, we have to extract some config(s) from the Consumer. This is a pretty awkward hack, but now that I've thought it through, it seems the best we can do. Maybe we can mull it over and file an improvement jira for the Consumer to improve use cases like this.

Anyway, it seems like the "poll time" config is irrelevant, we just need to know what config to grab that corresponds to completing a fetch request with high probability. It seems like we shouldn't need to update metadata, so we would send a fetch request on the first poll call, and we just need to block for whatever time bounds the fetch response time.

I'm honestly not sure what timeout would be best here. It looks like the ConsumerNetworkClient will just wait for a response until it gets a "disconnect" (L598). Is that a socket timeout? I'm not sure.

Reply:
I agree. The "issue" is really that `poll()` does _not_ throw a `TimeoutException`... Also, because we do manual assignment, `poll()` would never return "early" as it never need to wait for joining a consumer group. -- However, compare to `max.task.idle.ms`, we are in a better situation here, because we poll() for only a single partition at a time.

I also agree, that applying `task.timeout.ms` should start _after_ we got a first timeout -- this was how the original code worked that you criticized as:
> Man, this is confusing.

And I agree, that the code was not straightforward to understand. But if we think it's the right thing to do, I am also happy to add it back :)

I am also not an expert on all consumer internals, but from my understanding, fetch requests are send async in general, and if a fetch request fails, the consumer would actually not retry it but a retry would be triggered by the next `poll()` call. If there is no data available (ie, fetch request did not return yet) when `poll()` is called, the consumer would block internally until `poll(Duration)` timeout expires or until a fetch request returns (whatever comes first).

Furthermore, before `poll()` returns, it always check if a fetch request is in-flight or not, and sends one if not.

Thus, on the verify first call to `poll()` we know that no fetch request can be in-flight and we also know that `poll()` would send one, and block until it returns or `poll(Duration)` expired. Thus, if `poll()` does not block for at least `request.timeout.ms`, and we get empty back we don't know which case holds, however, if we use the request timeout, it seems that we _know_ if the fetch was successful or did time out? We also know, that a fetch request will be inflight after `poll()` returns. Thus, for any consecutive `poll()` applying request timeout also ensures that we know if the request was successful or not.

I guess the only difference to what I just described to my original code was, that I uses `pollTime + requestTimeout`.

Bottom line: I am not 100% sure what you propose? Should we go with the original design? Or with the new design? -- In the end, I think we don't need a follow up PR, and we can just try to get it right in this PR. I don't see any benefit in splitting it up into 2 PRs (because, as mentioned above, we fetch for a single partitions and thus it's a different case compared to `max.task.idle.ms` scenario).

Reply:
Ok, thanks @mjsax . I just traced though the consumer code again, and have finally been able to see what you already knew: that `request.timeout.ms` is indeed the correct amount of time to wait.

Namely, we send a fetch here: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L1285

Which calls through to client.send here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java#L263

Which fills in the `request.timeout.ms` config value here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java#L106

Which uses it to construct a ClientRequest here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java#L129-L130

Which then gets used to create an InFlightRequest when it gets sent here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L1248

Which is later used to detect expired requests here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java#L162

Which is used to list nodes (brokers) for which there is an expired request here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java#L179

Which is then processed as a "disconnection" here: https://github.com/apache/kafka/blob/659ca8f089c6b1c3a53036a8f96de7f763971fd1/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L803

It also looks like the KafkaClient just does a tight-loop checking for a network response, so we don't really need any extra time to account for sampling errors. Also, it still seems like using the sum as the poll duration is just as good as using your retry logic, so I think the duration parameter is fine.

My only remaining question, which maybe doesn't really matter one way or another, is whether `poll.ms` really belongs here or not. It seems like the desired semantics are accomplished by just waiting `request.timeout.ms` for the initial failure, and then an extra `task.timeout.ms` for any retries.

Reply:
Thanks for digging into it @vvcephei -- The question about `pollTimeout` is fair. I guess for default configs it might not matter much, as the default is 100ms IIRC.

At the same time, we apply the same `pollTimeout` during regular processing, and as a matter of fact, for this case, a use might want to use a longer poll-timeout, as otherwise, the thread would just "busy wait" anyway (only the responsiveness for a shutdown of the app should be considered).

Thus, it might actually make sense to exclude `pollTimeout` completely and only use `requestTimeout + taskTimeout`. Again, using `taskTimeout` in `poll()` reduces the responsiveness of a shutdown -- however, atm during bootstrapping we ignore a shutdown signal anyway, hence, for now we don't make the situation worse. I create a ticket to fix this: https://issues.apache.org/jira/browse/KAFKA-10317 and will add a comment for now.

Short related note: actually using `requestTimeout` seems to be a conservative upper bound for poll(). A request could fail with a different error before `requestTimeout` hits and would be retried internally for this case -- if this happens, we might want to start the `taskTimeout` earlier. However, we don't have any means atm to detect this case. Thus, using `requestTimeout` is the best option we have right now (because triggering `taskTimeout` too early seems to be worse than triggering it too late). I created a ticket though that might allow us to improve the code later: https://issues.apache.org/jira/browse/KAFKA-10315